### ISSUE TITLE
（很急）运行DeepFM以及往其他模型中添加FM时报错ValueError: Dimension 0 in both shapes must be equal, but are 1 and 64.该如何解决

### ISSUE THREAD

**Issue by Motsen** [shenweichen/DeepCTR#448]
**Describe the question(问题描述)**
运行DeepFM以及往其他模型中添加FM时报错ValueError: Dimension 0 in both shapes must be equal, but are 1 and 64. Shapes are [1] and [64]. for '{{node functional_1/concatenate_1/concat_1}} = ConcatV2[N=25, T=DT_BOOL, Tidx=DT_INT32](functional_1/concatenate_1/ones_like, functional_1/concatenate_1/ones_like_1, functional_1/concatenate_1/ones_like_2, functional_1/concatenate_1/ones_like_3, functional_1/concatenate_1/ones_like_4, functional_1/concatenate_1/ones_like_5, functional_1/concatenate_1/ones_like_6, functional_1/concatenate_1/ones_like_7, functional_1/concatenate_1/ones_like_8, functional_1/concatenate_1/ones_like_9, functional_1/concatenate_1/ones_like_10, functional_1/concatenate_1/ones_like_11, functional_1/concatenate_1/ones_like_12, functional_1/concatenate_1/ExpandDims, functional_1/concatenate_1/ones_like_13, functional_1/concatenate_1/ones_like_14, functional_1/concatenate_1/ones_like_15, functional_1/concatenate_1/ones_like_16, functional_1/concatenate_1/ones_like_17, functional_1/concatenate_1/ones_like_18, functional_1/concatenate_1/ones_like_19, functional_1/concatenate_1/ones_like_20, functional_1/concatenate_1/ones_like_21, functional_1/concatenate_1/ones_like_22, functional_1/concatenate_1/ones_like_23, functional_1/concatenate_1/concat_1/axis)' with input shapes: [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,1], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [?,1,64], [] and with computed input tensors: input[25] = <1>.

**Additional context**
[deepfm.zip](https://github.com/shenweichen/DeepCTR/files/8162443/deepfm.zip)
运行代码后报错

**Operating environment(运行环境):**
python version 3.9
tensorflow version 2.3.0
deepctr version 0.8.7

**Comment by jexterliangsufe** [comment]
因为FM层的输入维度是N * P * H，N是batch_size，P是离散特征个数（也可以说是SparseFeature），H是embedding的维度，而你输入FM层的数据维度的H不是固定的，不同的特征采用了不同的H，所以会报错，我也遇到了这个问题，目前的解决方法是把所有SparseFeature的embedding维度设置成一样的，不知道现在你有没有更好的解决方法呢？
FM layer's inout dimension is N * P * H. N represents the batch size, P is the number of sparse feautures and H is the dimension of embeddings. However, your H of FM inputs is mixed. In other words, you use different embedding dims for different sparse features. It is the direct cause of your error. I met this problem and my solution is setting the embedding dims of sparse features to be same. 
You may reply me if you have better solutions now~

**Comment by emremrah** [comment]
> 因为FM层的输入维度是N * P * H，N是batch_size，P是离散特征个数（也可以说是SparseFeature），H是embedding的维度，而你输入FM层的数据维度的H不是固定的，不同的特征采用了不同的H，所以会报错，我也遇到了这个问题，目前的解决方法是把所有SparseFeature的embedding维度设置成一样的，不知道现在你有没有更好的解决方法呢？ FM layer's inout dimension is N * P * H. N represents the batch size, P is the number of sparse feautures and H is the dimension of embeddings. However, your H of FM inputs is mixed. In other words, you use different embedding dims for different sparse features. It is the direct cause of your error. I met this problem and my solution is setting the embedding dims of sparse features to be same. You may reply me if you have better solutions now~


Hi. Can you explain why embedding dims of different sparse features must be the same?

**Comment by zanshuxun** [comment]
> > 因为FM层的输入维度是N * P * H，N是batch_size，P是离散特征个数（也可以说是SparseFeature），H是embedding的维度，而你输入FM层的数据维度的H不是固定的，不同的特征采用了不同的H，所以会报错，我也遇到了这个问题，目前的解决方法是把所有SparseFeature的embedding维度设置成一样的，不知道现在你有没有更好的解决方法呢？ FM layer's inout dimension is N * P * H. N represents the batch size, P is the number of sparse feautures and H is the dimension of embeddings. However, your H of FM inputs is mixed. In other words, you use different embedding dims for different sparse features. It is the direct cause of your error. I met this problem and my solution is setting the embedding dims of sparse features to be same. You may reply me if you have better solutions now~
> 
> Hi. Can you explain why embedding dims of different sparse features must be the same?

How to perform FM operation on two vectors with different dimensions?

**Comment by zhangjianzoujianghu18** [comment]
将from .layers.utils import concat_func 中的concat_func函数改为下面的函数，就可以运行了
只要是用Lambda(lambda x: tf.concat(x, axis=-1))(inputs) 替换 Concatenate(axis=axis)(inputs)
from tensorflow.keras.layers import  Lambda
def concat_func(inputs, axis=-1, mask=False):
    if not mask:
        inputs = list(map(NoMask(), inputs))
    if len(inputs) == 1:
        return inputs[0]
    else:
        return Lambda(lambda x: tf.concat(x, axis=-1))(inputs)
